src/main/kotlin/org/rust/lang/core/psi/ext/RsCompositeElement.kt

```diff
@@ -25,40 +24,36 @@ interface RsCompositeElement : PsiElement {
     val containingMod: RsMod
 }
 
-val RsCompositeElement.cargoWorkspace: CargoWorkspace? get() {
-    // It's important to look the module for `containingFile` file
-    // and not the element itself. Otherwise this will break for
-    // elements in libraries.
-    val module = ModuleUtilCore.findModuleForPsiElement(containingFile)
-    if (module != null) return module.cargoWorkspace
-
-    // The element is outside of a module. Most likely, it is an element
-    // from the library created by `RootsProvider`. Let's just hope there's
-    // a single workspace in this case and return it.
-    // Ideally, we need something more clever here, because there may be two
-    // workspaces, which share the same library.
-    return project.modules.map { it.cargoWorkspace }.firstOrNull { it != null }
-}
+val CARGO_WORKSPACE = Key.create<CargoWorkspace>("CARGO_WORKSPACE")
+val RsCompositeElement.cargoWorkspace: CargoWorkspace?
+    get() {
+        val psiFile = containingFile.originalFile
+        psiFile.getUserData(CARGO_WORKSPACE)?.let { return it }
+        val vFile = psiFile.virtualFile ?: return null
+        return project.cargoProjects.findProjectForFile(vFile)?.workspace
+    }
 
 
-val RsCompositeElement.crateRoot: RsMod? get() {
-    return if (this is RsFile) {
-        val root = superMods.lastOrNull()
-        if (root != null && root.isCrateRoot)
-            root
-        else
-            null
-    } else {
-        (context as? RsCompositeElement)?.crateRoot
+val RsCompositeElement.crateRoot: RsMod?
+    get() {
+        return if (this is RsFile) {
```

Comment:
=

Reply:
this extension feels like it should be just in RsCompositeElement as virtual method

Reply:
:+1: 💯